src/realm/sync/network/default_socket.cpp

```diff
@@ -371,10 +378,177 @@ void DefaultWebSocketImpl::initiate_websocket_handshake()
 }
 } // namespace
 
+///
+/// DefaultSocketProvider - default socket provider implementation
+///
+
+DefaultSocketProvider::DefaultSocketProvider(const std::shared_ptr<util::Logger>& logger,
+                                             const std::string user_agent)
+    : m_logger_ptr{logger}
+    , m_service{}
+    , m_random{}
+    , m_user_agent{user_agent}
+    , m_mutex{}
+    , m_state{State::Stopped}
+    , m_state_cv{}
+    , m_thread{}
+{
+    REALM_ASSERT(m_logger_ptr);                     // Make sure the logger is valid
+    util::seed_prng_nondeterministically(m_random); // Throws
+    start_keep_running_timer();                     // TODO: Update service so this timer is not needed
+    start();
+}
+
+DefaultSocketProvider::~DefaultSocketProvider()
+{
+    // Wait for the thread to stop
+    stop(true);
+    if (m_keep_running_timer)
+        m_keep_running_timer->cancel();
+    if (m_thread.joinable()) {
+        m_thread.join();
+    }
+    // Shutting down - no need to lock mutex before check
+    REALM_ASSERT(m_state == State::Stopped);
+}
+
+void DefaultSocketProvider::start()
+{
+    std::unique_lock<std::mutex> lock(m_mutex);
+    // Has the thread already been started or is running
+    if (m_state == State::Starting || m_state == State::Started || m_state == State::Running)
+        return; // early return
+
+    // If the thread has been previously run, make sure it has been joined first
+    if (m_state == State::Stopping || m_state == State::Stopped) {
+        if (m_thread.joinable()) {
+            lock.unlock();
+            m_thread.join();
+            lock.lock();
+        }
+    }
+
+    m_logger_ptr->trace("Default event loop: start()");
+    REALM_ASSERT(m_state == State::Stopped);
+    do_state_update(State::Starting);
+    m_thread = std::thread{make_event_loop()};
+    lock.unlock();
+    // Wait for the thread to start before continuing
+    state_wait_for(State::Started);
+}
+
+std::function<void()> DefaultSocketProvider::make_event_loop()
+{
+    return [this]() {
```

Comment:
If you want to simplify this a bit, you can actually just make a member function the body of the thread - something like this https://godbolt.org/z/Ge6WMr9hv

Reply:
Thanks - in `std::thread(&ThreadyThing::thread_body, this)` the `this` is what I was missing from trying to call a function directly... it makes sense, since the first (hidden) parameter for any class function is `this`.